### PR TITLE
pre-release 4 branch merge

### PR DESCRIPTION
**Pull request by o-smirnov**

### CODE REVIEW

**Comment by o-smirnov** [comment]
@Athanaseus, I made a 1.0-pre4 release (turns out you can release off a branch too, not just off master), could you please make sure it goes into PyPI? Once it's there, I can test the 1.0-pre4 client.

File: bin/bootstrap-radiopadre-install [hunk 1]
```diff
@@ -103,12 +105,17 @@ if not os.path.exists(activation_script):
...
+        extras = " ".join(options.venv_extras.split(","))
+        message(f"Installing specified extras: {extras}")
+        shell(f"pip install {extras}")
 
 message("Activating virtualenv")
 with open(activation_script) as f:
     code = compile(f.read(), activation_script, 'exec')
     exec(code, dict(__file__=activation_script), {})
-
+
```
Comment (Athanaseus) [inline bin/bootstrap-radiopadre-install]: some white space over here

File: bin/setup-radiopadre-virtualenv [hunk 2]
```diff
@@ -106,45 +123,59 @@ for line in out.decode().split("\n"):
...
+shutil.copy2(PADRE_PATH + "/html/radiopadre-kernel.js", kerneldir + "/kernel.js")
 
 # add link to radiopadre web components in notebook static dir
 notebook_padre_www = notebook_dir + "/notebook/static/radiopadre-www"
 padre_www = PADRE_PATH + "/html"
-message("making link from {notebook_padre_www} to {padre_www}")
-if not os.path.exists(notebook_padre_www):
+remove_installation(notebook_padre_www)
+if options.editable:
+    message("making link from {notebook_padre_www} to {padre_www}")
```
Comment (Athanaseus) [inline bin/setup-radiopadre-virtualenv]: missing formating (f-string)

File: bin/setup-radiopadre-virtualenv [hunk 3]
```diff
@@ -106,45 +123,59 @@ for line in out.decode().split("\n"):
...
     os.symlink(padre_www, notebook_padre_www)
+else:
+    message("copying {notebook_padre_www} to {padre_www}")
```
Comment (Athanaseus) [inline bin/setup-radiopadre-virtualenv]: ditto

File: bin/setup-radiopadre-virtualenv [hunk 4]
```diff
@@ -212,15 +243,26 @@ if not options.no_js9:
 
     # Make symlink to js9 install dir in notebook dir
     notebook_js9_www = notebook_dir + "/notebook/static/js9-www"
-    message("making link from {notebook_js9_www} to {js9_www}")
-    if not os.path.exists(notebook_js9_www):
+    
+    remove_installation(notebook_js9_www)
+    if options.editable:
+        message("making link from {notebook_js9_www} to {js9_www}")
```
Comment (Athanaseus) [inline bin/setup-radiopadre-virtualenv]: ditto

File: bin/setup-radiopadre-virtualenv [hunk 5]
```diff
@@ -212,15 +243,26 @@ if not options.no_js9:
...
         os.symlink(js9_www, notebook_js9_www)
+    else:
+        message("copying {js9_www} to {notebook_js9_www}")
```
Comment (Athanaseus) [inline bin/setup-radiopadre-virtualenv]: ditto

File: bin/setup-radiopadre-virtualenv [hunk 6]
```diff
@@ -212,15 +243,26 @@ if not options.no_js9:
...
 
     notebook_socket_io = notebook_dir + "/notebook/static/socket.io.js"
     js9_socket_io = js9_www + "/node_modules/socket.io-client/dist/socket.io.js"
-    message("making link from {notebook_socket_io} to {js9_socket_io}")
-    if not os.path.exists(notebook_socket_io):
+    
+    remove_installation(notebook_socket_io)
+    
+    if options.editable:
+        message("making link from {notebook_socket_io} to {js9_socket_io}")
```
Comment (Athanaseus) [inline bin/setup-radiopadre-virtualenv]: ditto

File: bin/setup-radiopadre-virtualenv [hunk 7]
```diff
@@ -212,15 +243,26 @@ if not options.no_js9:
...
         os.symlink(js9_socket_io, notebook_socket_io)
+    else:
+        message("copying {js9_socket_io} to {notebook_socket_io}")
```
Comment (Athanaseus) [inline bin/setup-radiopadre-virtualenv]: ditto

**Comment by Athanaseus** [comment]
> @Athanaseus, I made a 1.0-pre4 release (turns out you can release off a branch too, not just off master), could you please make sure it goes into PyPI? Once it's there, I can test the 1.0-pre4 client.

Awesome! Yep, that went through.
PS: I'll do a test install

File: bin/setup-radiopadre-virtualenv (hunk 2 above)
Comment (Athanaseus) [reply bin/setup-radiopadre-virtualenv]: oh i see the message `func` handles this

Comment (o-smirnov) [reply bin/setup-radiopadre-virtualenv]: Yeah that was me inventing f-strings before there were f-strings...

**Comment by Athanaseus** [comment]
The installation seems to work for me!

**Comment by o-smirnov** [comment]
Excellent! I'm still having some trouble with remote deployments, but I have to work on something else now. Will get back to this over the next day or two.

**Pull request merged**